### PR TITLE
dyndep: dynamically discovered dependencies for Fortran and C++20 modules

### PR DESCRIPTION
**Pull request by bradking**
Fortran modules (and likely C++20 modules) require sources that provide modules to be compiled before sources that consume modules.  Therefore the content of the source files defines dependencies.  Unlike preprocessor dependencies (handled by `depfile`), module dependencies must be discovered *before* compiling sources even on the first build.  This PR introduces a `dyndep` feature to provide buildsystem generators with a way to handle such dynamically discovered dependencies.

For some history, see the ninja-build list thread [Adding Ninja features needed for Fortran](https://groups.google.com/d/topic/ninja-build/tPOcu5EWXio/discussion) and issue #1265.

This PR includes the full implementation, documentation, and thorough tests.  The changes have been used in practice by CMake users for years (via a Kitware-maintained branch of ninja).  They have been used to build large Fortran projects with thousands of sources and hundreds of modules.

The feature works by providing a new `dyndep = ` binding for build statements.  It specifies one of the (order-only) inputs that will provide the dynamically discovered dependencies that must be loaded before the build statement's command is executed.  Once the "dyndep" file is ready, ninja loads it to inject additional dependencies in its build graph, re-evaluate the affected portion of the graph, and then continue the build normally.

For example:

```
rule f95
  command = f95 -o $out -c $in
rule fscan
  command = fscan -o $out $in

build foobar.dd: fscan foo.f90 bar.f90

build foo.o: f95 foo.f90 || foobar.dd
  dyndep = foobar.dd
build bar.o: f95 bar.f90 || foobar.dd
  dyndep = foobar.dd
```

In this example the order-only dependencies ensure that `foobar.dd` is generated before either source compiles.  The hypothetical `fscan` tool (provided by the buildsystem generator in practice) scans the source files and writes `foobar.dd` with content such as:

```
ninja_dyndep_version = 1
build foo.o | foo.mod: dyndep
build bar.o: dyndep |  foo.mod
```

The "dyndep" file specifies additional inputs and outputs for the relevant build statements beyond those specified in the build manifest.

See documentation added by this PR for more details and examples.

Fixes: #1265

### CODE REVIEW

**Comment by mqudsi** [comment]
Please forgive me if I'm overstepping here, it seems like a lot of good work has gone into this - but is it necessary to required `foobar.dd` to be listed both as the `dyndep` source and explicitly as an order-only dependency? Can't ninja deduce the latter?

**Comment by bradking** [comment]
@mqudsi I did think about that while designing this but decided against having dependencies specified in `build.ninja` by means other than the normal syntax.  Also, the dyndep input does not have to be an order-only input and the binding would not have a clear way to specify which type.  Currently it is an error for the `dyndep = ` binding to name a path that is not an input so we would have an option in the future to replace the error case with an implied input.

**Comment by bradking** [comment]
Rebased on `master` for post-1.9 development.

**Comment by dublet** [comment]
@bradking Is there any CMake branch where this feature is used?

**Comment by bradking** [comment]
@dublet CMake 3.7 and higher already have support.  It is activated by recognizing a special `.dyndep-1` suffix on Ninja's version number that Kitware added to our branch of Ninja we maintain pending upstream integration.  One can activate it on top of this PR's branch by locally hacking `src/version.cc`:

```diff
-const char* kNinjaVersion = "1.9.0.git";
+const char* kNinjaVersion = "1.9.0.git.dyndep-1";
```

Of course once Ninja integrates this PR in a release then we can update CMake to check for that instead.

**Comment by mathstuf** [comment]
This is probably future work, but outputting a build graph with dyndep information can help to find bottlenecks in the build. Is this already available with this PR or should I open an issue for it?

**Comment by bradking** [comment]
@mathstuf I updated the branch here to load dyndep files for the clean, graph, and query tools.

**Comment by bradking** [comment]
@jhasse thanks for the review.  I've updated the branch to address all review comments, either with fixes or additional source comments with answers to your questions.

File: src/build_test.cc [hunk 1]
```diff
@@ -602,19 +648,33 @@ bool FakeCommandRunner::WaitForCommand(Result* result) {
...
-  last_command_ = NULL;
+
+  // Provide a way for test cases to verify when an edge finishes that
+  // some other edge is still active.  This is useful for test cases
+  // covering behavior involving multiple active edges.
+  string const& verify_active_edge = edge->GetBinding("verify_active_edge");
+  if (!verify_active_edge.empty()) {
+    bool verify_active_edge_found = false;
+    for (vector<Edge*>::iterator i = active_edges_.begin(),
+         e = active_edges_.end(); !verify_active_edge_found && i != e; ++i) {
```
Comment (jhasse) [inline src/build_test.cc]: I would remove `!verify_active_edge_found && ` and use break below (maybe even skip that as this is only used in one test and not performance critical). While creating `e` is faster than calling `end()` in the condition, I wouldn't use that trick here as it makes the loop more complicated as it needs to be.

File: src/build_test.cc [hunk 2]
```diff
@@ -602,19 +648,33 @@ bool FakeCommandRunner::WaitForCommand(Result* result) {
     result->status = ExitFailure;
   else
     result->status = ExitSuccess;
-  last_command_ = NULL;
+
+  // Provide a way for test cases to verify when an edge finishes that
+  // some other edge is still active.  This is useful for test cases
+  // covering behavior involving multiple active edges.
+  string const& verify_active_edge = edge->GetBinding("verify_active_edge");
```
Comment (jhasse) [inline src/build_test.cc]: Please use west-const.

**Comment by bradking** [comment]
@jhasse thanks for the review. I've updated the branch to address all review comments and rebase on `master`.

**Comment by jhasse** [comment]
Awesome stuff! Again: Thank you so much for the hard work :)

**Comment by bradking** [comment]
@jhasse great, thanks for the review and integration!

**Comment by Ericson2314** [comment]
https://github.com/ninja-build/ninja/issues/1303 Per this long thread, I think it would be great to have an *optional* warning for `depfile` saying to use `dyndep` instead. I'm really stoked for `dyndep`; I would always like to use it instead of `depfile` so initial builds and rebuilds do the same thing.

**Comment by melak47** [comment]
In general, how is the scanner supposed to know the name of the *output* file?
Say different source files produce their output in different directories, like
```
build bob/foo.o: f95 foo.f90
build alice/bar.o: f95 bar.f90
```
Now, generating a single `foobar.dd` with a single invocation of `fscan` would be difficult, the generator would have to pass the output file names for each input file into the scanner.
Then the scanner writes that information back out to the dyndep file, then ninja parses it back in.

Is that how it's meant to work?

**Comment by Ericson2314** [comment]
@melak47 I think it's basically left to the user to ensure the dyndeps file contains the needed information. You can use less or more dyndeps as is convenient; more would help with not wanted to gather up everything for one big scan.

**Comment by bradking** [comment]
> the generator would have to pass the output file names for each input file into the scanner....Is that how it's meant to work?

Yes.  For example, CMake's Fortran build rules pass `--obj=...` to the scanner to tell it what object file will be produced by compiling the source it is scanning.

**Pull request merged**